### PR TITLE
Solution to Enumerables Project

### PR DESCRIPTION
**Pull request by Jhonatanssal**
Building an Enumerable module with the next methods:

#my_each
#my_each_with_index
#my_select
#my_all?
#my_any?
#my_none?
#my_count
#my_map
#my_inject
We also modified #my_map in order to accomplish the requirements to accept blocks and Procs

Happy coding.

### CODE REVIEW

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  def my_each
+    var = to_a
+    e = size - 1
+    return unless block_given?
+
+    0.upto(e) do |x|
+      yield(var[x])
+    end
+    var
+  end
```
Comment (BrittanyBlake) [inline enumerable.rb]: There are two issues that need to me solved for your `my_each` method:

- [x] First, `my_each` should return an enumerator if no block is given. 

- [x] If a block is given and self is a range, it should return the range itself, after calling the given block once for each element in self. Yours returns all of the elements in the range.

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  def my_each_with_index
+    var = to_a
+    e = size - 1
+    return unless block_given?
+
+    0.upto(e) do |x|
+      yield(var[x], x)
+    end
+    var
+  end
```
Comment (BrittanyBlake) [inline enumerable.rb]: There are also two issues that need to be solved in `my_each_with_index`:

- [x] Like above(`my_each`), `my_each_with_index`should also return an enumerator if no block is given.

- [x] The second issue is also the same as above in `my_each`. When a block is given and self is a range, `my_each_with_index` should return the range itself after calling the given block once for each element in self.

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    var = to_a
+    e = size - 1
+    arr = []
+    if block_given?
+      0.upto(e) do |x|
+        arr.push(var[x]) if yield(var[x])
+      end
+    end
+    arr
+  end
```
Comment (BrittanyBlake) [inline enumerable.rb]: - [x] Like with the previous two methods, `my_select` should also return an enumerator if no block is given.

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+      var.my_each do |x|
+        arr.push(yield(x))
+      end
+    else
+      var.my_each do |x|
+        arr.push(proc[x]) if proc.is_a?(Proc)
+      end
+    end
+    arr
+  end
```
Comment (BrittanyBlake) [inline enumerable.rb]: There are two issues with `my_map` that need to be addressed: 
- [x] `my_map` should also return an enumerator if no block is given.

- [x] If both a block and a proc are given, `my_map` should only execute the proc and return a new array with the results of running the proc once for every element.

File: enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+      end
+      acc
+    elsif arg.size == 2
+      acc = arg[0]
+      0.upto(b) do |i|
+        acc = acc.send(arg[1], var[i])
+      end
+      acc
+    end
+  end
```
Comment (BrittanyBlake) [inline enumerable.rb]: - [x] `my_inject` should raise a `LocalJumpError` when no block or argument is given. You are not currently raising any error for this method.

**Review by BrittanyBlake — CHANGES_REQUESTED** [review]
### Status: Changes Required ♻️ 
****
### Hi @jssarrazolaa  and @ABDELLANI-Youcef ! 👋 
 You have also done a wonderful job with this difficult project so far! The majority of your methods are working as they should 👏 👏 👏  There are just a few issues that need to be resolved before you can move on. You can find a summary of the issues below ⬇️ , but please check the comments for more details.
****

-  A few of your methods should are not currently returning an enumerator when no block is given.

-  Both your `my_each` and `my_each_with_index` methods need to be fixed when working with ranges. 

-  Your `my_map` method is not currently handling a scenario when both a block and a proc are given well.

-  Your `my_inject` should raise an error when no block is given. 

****
### Please make the required changes and resubmit for a new code review once you are ready 🙏

Good Luck and Happy Coding 🍀 👩🏾‍💻

_P.S. If you have any questions about my review or need help, you can message me on slack (@ Brittany Blake)._

**Review by tirthajyoti-ghosh — APPROVED** [review]
# Approved ✅

## You have done an incredible job making this project 👏🧨💪! Just perfect 💯! 

### You have met all the requirements of the project. Go ahead and merge it! Onto the next project 🚀 !

Happy coding 👨‍💻 

<hr>

**Tirthajyoti Ghosh**

[GitHub](https://github.com/tirthajyoti-ghosh) | [LinkedIn](https://www.linkedin.com/in/tirthajyoti-ghosh/) | [Twitter](https://twitter.com/terrific_ghosh)

**Pull request merged**